eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs

```diff
@@ -242,34 +247,57 @@ instance
       k -> invalidKey k
 
 delegationTransition ::
-  (ShelleyEraAccounts era, ShelleyEraTxCert era, EraPParams era, AtMostEra "Babbage" era) =>
+  ( EraCertState era
+  , ShelleyEraAccounts era
+  , ShelleyEraTxCert era
+  , EraPParams era
+  , AtMostEra "Babbage" era
+  ) =>
   TransitionRule (ShelleyDELEG era)
 delegationTransition = do
-  TRC (DelegEnv slot epochNo ptr chainAccountState pp, ds, c) <- judgmentContext
+  TRC (DelegEnv slot epochNo ptr chainAccountState pp, certState, c) <- judgmentContext
   let pv = pp ^. ppProtocolVersionL
+      ds = certState ^. certDStateL
   case c of
     RegTxCert cred -> do
       -- (hk ∉ dom (rewards ds))
       not (isAccountRegistered cred (ds ^. accountsL)) ?! StakeKeyAlreadyRegisteredDELEG cred
       let compactDeposit = compactCoinOrError (pp ^. ppKeyDepositL)
-      pure (ds & accountsL %~ registerShelleyAccount cred ptr compactDeposit Nothing)
+      pure $ certState & certDStateL . accountsL %~ registerShelleyAccount cred ptr compactDeposit Nothing
     UnRegTxCert cred -> do
       let !(mAccountState, !accounts) = unregisterShelleyAccount cred (ds ^. accountsL)
           checkStakeKeyHasZeroRewardBalance = do
             accountState <- mAccountState
             let accountBalance = accountState ^. balanceAccountStateL
             guard (accountBalance /= mempty)
             Just $ fromCompact accountBalance
-      -- (hk ∈ dom (rewards ds))
-      isJust mAccountState ?! StakeKeyNotRegisteredDELEG cred
       failOnJust checkStakeKeyHasZeroRewardBalance StakeKeyNonZeroAccountBalanceDELEG
-      pure $ ds & accountsL .~ accounts
+      -- (hk ∈ dom (rewards ds))
+      case mAccountState of
+        Nothing -> do
+          failBecause $ StakeKeyNotRegisteredDELEG cred
+          pure certState
+        Just accountState ->
+          pure $
+            certState
+              & certDStateL . accountsL .~ accounts
+              & certPStateL
+                %~ unDelegStakePool cred (accountState ^. stakePoolDelegationAccountStateL) Nothing
     DelegStakeTxCert cred stakePool -> do
       -- note that pattern match is used instead of cwitness and dpool, as in the spec
       -- (hk ∈ dom (rewards ds))
-      isAccountRegistered cred (ds ^. accountsL) ?! StakeDelegationImpossibleDELEG cred
-      pure $
-        ds & accountsL %~ adjustAccountState (stakePoolDelegationAccountStateL ?~ stakePool) cred
+      case lookupAccountState cred (ds ^. accountsL) of
+        Nothing -> do
+          failBecause $ StakeDelegationImpossibleDELEG cred
+          pure certState
+        Just currentAccount ->
```

Comment:
There is no other type of account state except current, so I think it makes sense to drop the prefix
```suggestion
        Just accountState ->
```